tests/panel_test.js

```diff
@@ -0,0 +1,565 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+/**
+ * @fileoverview This suite of tests verifies that the graph is correctly
+ *     constructed and styled.
+ */
+goog.provide('audion.entryPoints.PanelTest');
+
+goog.require('audion.entryPoints.panel');
+goog.require('audion.messaging.MessageType');
+goog.require('goog.testing.PropertyReplacer');
+goog.require('goog.testing.jsunit');
+
+// Used to temporarily replace properties.
+var stubs;
+
+
+var setUp = function() {
+  stubs = new goog.testing.PropertyReplacer();
+
+  // Disable requestAnimationFrame for now. It would add asynchronicity to this
+  // test, which could flake tests.
+  stubs.set(goog.global, 'requestAnimationFrame', goog.nullFunction);
+
+  // Reset the panel as if the user just visited a new page.
+  audion.entryPoints.resetPanel();
+};
+
+
+var tearDown = function() {
+  // Bring back any replaced properties.
+  stubs.reset();
+};
+
+
+var testNewNodesAndEdges = function() {
+  // The user creates a destination node in frame 42 with ID 3.
+  goog.global.acceptMessage({
+    type: audion.messaging.MessageType.NODE_CREATED,
+    frameId: 42,
+    isOffline: false,
+    nodeId: 3,
+    nodeType: 'AudioDestination',
+    numberOfInputs: 1,
+    numberOfOutputs: 1,
+    audioParamNames: [],
+  });
+
+  // The user creates a gain node in the same frame with ID 4.
+  goog.global.acceptMessage({
+    type: audion.messaging.MessageType.NODE_CREATED,
+    frameId: 42,
+    isOffline: false,
+    nodeId: 4,
+    nodeType: 'Gain',
+    numberOfInputs: 1,
+    numberOfOutputs: 1,
+    audioParamNames: ['gain'],
+  });
+
+  // And then connects the gain node to the destination.
+  goog.global.acceptMessage({
+    type: audion.messaging.MessageType.NODE_TO_NODE_CONNECTED,
+    frameId: 42,
+    sourceNodeId: 4, // The gain node.
+    destinationNodeId: 3, // The destination node.
+    fromChannel: 0,
+    toChannel: 0
+  });
+
+  // The user makes an oscillator node with ID 5.
+  goog.global.acceptMessage({
+    type: audion.messaging.MessageType.NODE_CREATED,
+    frameId: 42,
+    isOffline: false,
+    nodeId: 5,
+    nodeType: 'Oscillator',
+    numberOfInputs: 1,
+    numberOfOutputs: 1,
+    audioParamNames: ['detune', 'frequency'],
+  });
+
+  // And connects the oscillator to the gain *AudioParam* of gain node 4.
+  goog.global.acceptMessage({
+    type: audion.messaging.MessageType.NODE_TO_PARAM_CONNECTED,
+    frameId: 42,
+    sourceNodeId: 5, // The oscillator node.
+    destinationNodeId: 4, // The gain node.
+    destinationParamName: 'gain',
+    fromChannel: 0
+  });
+
+  // Validate that the panel produces the correct graph. And styles + colors it.
+  assertObjectEquals({
```

Comment:
I wonder if this is too specific. Maybe we should assert for specifics like node type and edges present instead of comparing objects.